src/self_driving/planning/pilot_util.cpp

```diff
@@ -419,4 +425,233 @@ void PilotUtil::GroupFeaturesByOU(
+      "having max(stanumrows) > 0;");
+
+  common::Future<task::DummyResult> sync;
+  task_manager->AddTask(std::make_unique<task::TaskDML>(catalog::INVALID_DATABASE_OID, query,
+                                                        std::make_unique<optimizer::TrivialCostModel>(), false,
+                                                        to_row_fn, common::ManagedPointer(&sync)));
+
+  auto future_result = sync.WaitFor(Pilot::FUTURE_TIMEOUT);
+  if (!future_result.has_value()) {
+    throw PILOT_EXCEPTION("Future timed out.", common::ErrorCode::ERRCODE_IO_ERROR);
+  }
+  UNUSED_ATTRIBUTE bool success = future_result->second;
+  NOISEPAGE_ASSERT(success, "Failed reading from pg_statistic");
+
+  // Calculating queries' changes in the number of rows
+  auto &metadata = forecast->GetWorkloadMetadata();
+  auto txn = txn_manager->BeginTransaction();
+  // <query_id, <table_id, number of rows changed by that query in that table>>
+  std::unordered_map<execution::query_id_t, std::pair<catalog::table_oid_t, int64_t>> query_row_changes;
+  for (const auto &[query_id, query_text] : metadata.query_id_to_text_) {
+    // Parse the query
+    auto parse_result = parser::PostgresParser::BuildParseTree(query_text);
+    if (parse_result->NumStatements() == 0) continue;
+    auto statement = parse_result->GetStatement(0);
+
+    // Identify the table and the number of inserts/deletes
+    int64_t num_row_delta = 0;
+    catalog::table_oid_t table_oid;
+    auto accessor = catalog->GetAccessor(common::ManagedPointer(txn),
+                                         catalog::db_oid_t(forecast->GetDboidByQid(query_id)), DISABLED);
+    if (statement->GetType() == parser::StatementType::INSERT) {
+      auto insert_statement = reinterpret_cast<parser::InsertStatement *>(statement.Get());
+      table_oid = accessor->GetTableOid(insert_statement->GetInsertionTable()->GetTableName());
+      // Number of values is the number of rows inserted
+      num_row_delta = insert_statement->GetValues()->size();
+    }
+
+    if (statement->GetType() == parser::StatementType::DELETE) {
+      auto delete_statement = reinterpret_cast<parser::DeleteStatement *>(statement.Get());
+      table_oid = accessor->GetTableOid(delete_statement->GetDeletionTable()->GetTableName());
+      if (delete_statement->GetDeleteCondition() != nullptr) {
+        // TODO(lin): right now assuming only deleting one row when there's a condition. Need cardinality estimation
+        //  to get a more accurate estimation
+        num_row_delta = -1;
+      } else {
+        if (table_sizes.find(table_oid) != table_sizes.end())
+          // Delete all the tuples when without condition
+          num_row_delta = -table_sizes[table_oid];
+      }
+    }
+    if (num_row_delta != 0 && table_sizes.find(table_oid) != table_sizes.end())
+      query_row_changes[query_id] = std::make_pair(table_oid, num_row_delta);
+  }
+  txn_manager->Abort(txn);
```

Comment:
Commit should be cheap for read-only transactions (no logs to flush).  What's the reason for aborting instead?